### PR TITLE
feat(workflows): use built images in Github workflows

### PR DESCRIPTION
**Pull request by CarterFendley**
### Description of your changes:

This PR is an MVP which gets the images built and deployments patched to use those images. **The PR is intended to be iterated on with more formalization / improvement in the future** while also getting testing working in the meantime. This addition is related to #11239, but also makes changes related to `apiserver`, `persistenceagent`, and `scheduledworkflow` images, more discussion will be added to that issue.

The PR primarily modifies the scripts in `scripts/deploy/github/` which are used by the [GitHub action here](https://github.com/CarterFendley/pipelines/blob/627a3dbb7e5f0758877fb3b4681e59a88fe3b840/.github/actions/kfp-cluster/action.yml). The `kubectl patch` command is used to patch deployment images, and `kubectl set env` to modify environment variables.

**Other changes**

Some minor changes were made to the [wait_for_pods.py](https://github.com/CarterFendley/pipelines/blob/627a3dbb7e5f0758877fb3b4681e59a88fe3b840/scripts/deploy/github/kfp-readiness/wait_for_pods.py) tooling, to display more information in cases where there are errors. See the screenshot below of a case in which this helped me while working on this PR :P

Screenshot 2024-10-10 at 2 18 13 PM

### Checklist:
- [x] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

### Other comments

Since we are building more images, this does slow down the workflows a bit (`~17min` -> `~22min`), will try to think if we can speed that up a little because 22 minute waits are painful.

### CODE REVIEW

File: scripts/deploy/github/kfp-readiness/wait_for_pods.py [hunk 1]
```diff
@@ -41,8 +51,10 @@ def check_pods(calm_time=10, timeout=600, retries_after_ready=5):
         current_statuses = get_pod_statuses()
 
         logging.info("Checking pod statuses...")
-        for pod_name, (pod_status, ready, total) in current_statuses.items():
+        for pod_name, (pod_status, ready, total, waiting_messages) in current_statuses.items():
             logging.info(f"Pod {pod_name} - Status: {pod_status}, Ready: {ready}/{total}")
+            for waiting_msg  in waiting_messages:
+                logging.info(waiting_msg)
```
Comment (gmfrasca) [inline scripts/deploy/github/kfp-readiness/wait_for_pods.py:57]: Perhaps should provide a context message here before dumping the list of waiting messages, even if it's just a static "Reasons for Pending/Not Ready:" header line.  Going off of the example log posted it took a second to realize the `CONTAINER` lines were related to the previous pod status line before it

**Review by gmfrasca — COMMENTED** [review]
minor nitpick: update commit messages to use [Conventional commits syntax](https://www.conventionalcommits.org/en/v1.0.0/)

File: scripts/deploy/github/kfp-readiness/wait_for_pods.py [hunk 2]
```diff
@@ -21,15 +21,25 @@ def get_pod_statuses():
...
+        ready = 0
+        total = 0
+        waiting_messages = []
+        for status in container_statuses:
+            total += 1
+            if status.ready:
+                ready += 1
+            if status.state.waiting is not None:
+                if status.state.waiting.message is not None:
+                    waiting_messages.append(f'CONTAINER: {status.name} - {status.state.waiting.reason}: {status.state.waiting.message}')
```
Comment (gmfrasca) [inline scripts/deploy/github/kfp-readiness/wait_for_pods.py:33]: alternatively to https://github.com/kubeflow/pipelines/pull/11284/files#r1795992913, perhaps we can add additional context to `CONTAINER` prefix such as `Waiting on Container {name} - {reason}: {message}`

Comment (CarterFendley) [reply scripts/deploy/github/kfp-readiness/wait_for_pods.py:33]: Sure will take this option.

File: scripts/deploy/github/kfp-readiness/wait_for_pods.py (hunk 1 above)
Comment (CarterFendley) [reply scripts/deploy/github/kfp-readiness/wait_for_pods.py:57]: Took the alternative in your comment below, thanks for the suggestion!

File: scripts/deploy/github/deploy-kfp.sh [hunk 3]
```diff
@@ -41,6 +41,18 @@ then
   exit 1
 fi
 
+echo "Patching deployments to use built docker images..."
+# Patch API server
+kubectl patch deployment ml-pipeline -p '{"spec": {"template": {"spec": {"containers": [{"name": "ml-pipeline-api-server", "image": "kind-registry:5000/apiserver"}]}}}}' -n kubeflow
+# Patch persistance agent
```
Comment (hbelmiro) [inline scripts/deploy/github/deploy-kfp.sh:47]: Typo...
```suggestion
# Patch persistence agent
```

**Review by hbelmiro — COMMENTED** [review]
Many thanks for working on this @CarterFendley.
You just forgot to sign off one commit and I left a nit comment about a typo.

/ok-to-test

File: scripts/deploy/github/deploy-kfp.sh (hunk 3 above)
Comment (CarterFendley) [reply scripts/deploy/github/deploy-kfp.sh:47]: Awesome, thank you!

**Review by hbelmiro — APPROVED** [review]
/lgtm

**Pull request merged**